cmd/metal-api/internal/metal/machine.go

```diff
@@ -157,10 +157,108 @@ type MachineNetwork struct {
+
+// NetworkType represents the type of a network
+type NetworkType struct {
+	Name           string `json:"name,omitempty"`
+	Private        bool   `json:"private,omitempty"`
+	PrivatePrimary bool   `json:"private_primary,omitempty"`
+	Shared         bool   `json:"shared,omitempty"`
+	Underlay       bool   `json:"underlay,omitempty"`
+	Supported      bool   `json:"-"`
+}
+
+var (
+	PrivatePrimaryUnshared NetworkType = NetworkType{
+		Name:           "privateprimaryunshared",
+		Private:        true,
+		PrivatePrimary: true,
+		Shared:         false,
+		Underlay:       false,
+		Supported:      true,
+	}
+	PrivatePrimaryShared NetworkType = NetworkType{
+		Name:           "privateprimaryshared",
+		Private:        true,
+		PrivatePrimary: true,
+		Shared:         true,
+		Underlay:       false,
+		Supported:      true,
+	}
+	PrivateSecondaryShared NetworkType = NetworkType{
+		Name:           "privatesecondaryshared",
+		Private:        true,
+		PrivatePrimary: false,
+		Shared:         true,
+		Underlay:       false,
+		Supported:      true,
+	}
+	// PrivateSecondaryUnshared this case is not a valid configuration
+	PrivateSecondaryUnshared NetworkType = NetworkType{
+		Name:           "privatesecondaryunshared",
+		Private:        true,
+		PrivatePrimary: false,
+		Shared:         false,
+		Underlay:       false,
+		Supported:      false,
+	}
+	External NetworkType = NetworkType{
+		Name:           "external",
+		Private:        false,
+		PrivatePrimary: false,
+		Shared:         false,
+		Underlay:       false,
+		Supported:      true,
+	}
+	Underlay NetworkType = NetworkType{
+		Name:           "underlay",
+		Private:        false,
+		PrivatePrimary: false,
+		Shared:         false,
+		Underlay:       true,
+		Supported:      true,
+	}
+	AllNetworkTypes []NetworkType = []NetworkType{PrivatePrimaryUnshared, PrivatePrimaryShared, PrivateSecondaryShared, PrivateSecondaryUnshared, External, Underlay}
+)
+
+// Is checks whether the machine network has the given type
+func (mn *MachineNetwork) Is(n NetworkType) bool {
+	return mn.Private == n.Private && mn.PrivatePrimary == n.PrivatePrimary && mn.Shared == n.Shared && mn.Underlay == n.Underlay
+}
+
+// NetworkType determines the network type based on the flags stored in the db entity.
+func (mn *MachineNetwork) NetworkType() (*NetworkType, error) {
+	var nt *NetworkType
+	for _, t := range AllNetworkTypes {
+		if mn.Is(t) {
+			nt = &t
+			break
+		}
+	}
+	if nt == nil {
+		return nil, fmt.Errorf("could not determine network type out of flags, underlay: %v, privateprimary: %v, private: %v, shared: %v", mn.Underlay, mn.PrivatePrimary, mn.Private, mn.Shared)
+	}
+
+	if nt.Supported {
+		return nt, nil
+	}
+	// This is for machineNetworks from a Allocation which was before NetworkType was introduced.
+	// We guess based on unset fields not present at this time and therefor are set to false.
+	if mn.Private && !mn.PrivatePrimary && !mn.Shared && !mn.Underlay {
+		return &PrivatePrimaryUnshared, nil
+	}
```

Comment:
So, this is actually code that needs to be removed in a future release, right? Can we mention this somewhere?